File: AgeOfArenaUnity/Assets/Scripts/TradingSystem.cs
using UnityEngine;

/// <summary>
/// Drives Trade Cart units (UnitType.TradeCart). Each cart finds the nearest
/// enemy-or-neutral Market on the map, walks to it, and returns to the player's
/// own Market to deposit a gold bonus proportional to the distance travelled.
/// One cart produces roughly 10–30 gold per round-trip depending on map distance.
/// Ticked from GameManager.
/// </summary>
public class TradingSystem : MonoBehaviour
{
    const float TradeGoldPerUnit = 0.18f; // gold earned per world-unit of travel
    const float MinGold          = 8f;
    const float DepositRange     = 4f;

    enum TradeState { SeekingMarket, Travelling, Returning }

    public void Tick(System.Collections.Generic.List<UnitEntity> units,
                     System.Collections.Generic.List<BuildingEntity> buildings,
                     float dt)
    {
        for (int i = 0; i < units.Count; i++)
        {
            var u = units[i];
            if (u == null || u.type != UnitType.TradeCart) continue;
            StepCart(u, buildings);
        }
    }

    void StepCart(UnitEntity u, System.Collections.Generic.List<BuildingEntity> buildings)
    {
        // Trade state in patrolA (home market) / patrolB (target market) + tradeActive /
        // tradeReturning. Driven entirely here (NOT the patrol auto-flip), so the cart is
        // paid exactly once per ROUND TRIP — on arrival back at the home market (AoE2),
        // not on the outbound leg, and never twice.
        if (!u.tradeActive)
        {
            // Phase 1: find home + target market, head out to the target.
            BuildingEntity home   = NearestMarket(buildings, u.transform.position, u.teamId,  true);
            BuildingEntity target = NearestMarket(buildings, u.transform.position, u.teamId, false);
            if (home == null || target == null) return;
            u.patrolA        = home.transform.position;
            u.patrolB        = target.transform.position;
            u.tradeActive    = true;
            u.tradeReturning = false;
            u.MoveTo(u.patrolB);
            return;
        }

        if (!u.tradeReturning)
        {
            // Phase 2 (outbound): reached the target market? pick up goods, head home.
            if (Vector3.Distance(u.transform.position, u.patrolB) < DepositRange)
            {
                u.tradeReturning = true;
                u.MoveTo(u.patrolA);
            }
            return;
        }

        // Phase 3 (inbound): reached the home market? deposit the gold and loop.
        if (Vector3.Distance(u.transform.position, u.patrolA) < DepositRange)
        {
            float tripDist = Vector3.Distance(u.patrolA, u.patrolB);
            float earnedF  = Mathf.Max(MinGold, tripDist * TradeGoldPerUnit);
            var   gm       = GameManager.Instance;
            if (gm != null)
            {
                // CARA: Market Caravan tech boosts trade-cart yield (×1.5). The same
                // distance-based route also models a Dock-based Trade Cog (water route)
                // — naval trade reuses this StepCart logic once a water map ships.
                earnedF *= gm.teamTech[u.teamId].TradeGoldMult;
                gm.teamRes[u.teamId].Gain(ResourceKind.Gold, Mathf.RoundToInt(earnedF));
            }
            // Re-find markets next tick (handles a market being built/destroyed) and repeat.
            u.tradeActive = false;
        }
    }

    static BuildingEntity NearestMarket(System.Collections.Generic.List<BuildingEntity> buildings,
                                        Vector3 from, int teamId, bool ownTeam)
    {
        BuildingEntity best = null;
        float bestSq = float.MaxValue;
        for (int i = 0; i < buildings.Count; i++)
        {
            var b = buildings[i];
            if (b == null || b.type != BuildingType.Market || b.underConstruction) continue;
            if (ownTeam ? b.teamId != teamId : b.teamId == teamId) continue;
            float sq = (b.transform.position - from).sqrMagnitude;
            if (sq < bestSq) { bestSq = sq; best = b; }
        }
        return best;
    }
}

File: web/src/game/TradingSystem.ts
/**
 * TradingSystem.cs port — Trade Carts travel from the player's Market to an
 * enemy/neutral Market and back, earning gold proportional to trip distance.
 * AoE2 formula: max(8, dist × 0.18) gold per round-trip.
 */
import * as THREE from "three";
import { BuildingType, UnitType } from "../core/GameTypes";
import { ResourceKind } from "../core/GameTypes";
import type { Unit } from "./Unit";
import type { Building } from "./Building";
import type { ResourceManager } from "../core/ResourceManager";

const TRADE_GOLD_PER_UNIT = 0.18;
const MIN_GOLD = 8;
const DEPOSIT_RANGE = 4;

interface TradeRoute {
  homePos: THREE.Vector3;
  targetPos: THREE.Vector3;
  returning: boolean;
  active: boolean;
}

export class TradingSystem {
  private readonly _routes = new Map<Unit, TradeRoute>();

  tick(
    units: Unit[],
    buildings: Building[],
    teamRes: ResourceManager[],
    dt: number,
  ): void {
    for (const u of units) {
      if (!u.alive || u.unitType !== UnitType.TradeCart) continue;
      this._step(u, buildings, teamRes);
    }
  }

  private _step(u: Unit, buildings: Building[], teamRes: ResourceManager[]): void {
    let route = this._routes.get(u);

    if (!route || !route.active) {
      const home   = this._nearestMarket(buildings, u.pos, u.teamId, true);
      const target = this._nearestMarket(buildings, u.pos, u.teamId, false);
      if (!home || !target) return;
      route = { homePos: home.pos.clone(), targetPos: target.pos.clone(), returning: false, active: true };
      this._routes.set(u, route);
      u.moveTo(route.targetPos.clone());
      return;
    }

    if (!route.returning) {
      if (u.pos.distanceTo(route.targetPos) < DEPOSIT_RANGE) {
        route.returning = true;
        u.moveTo(route.homePos.clone());
      }
      return;
    }

    if (u.pos.distanceTo(route.homePos) < DEPOSIT_RANGE) {
      const tripDist = route.homePos.distanceTo(route.targetPos);
      const gold = Math.round(Math.max(MIN_GOLD, tripDist * TRADE_GOLD_PER_UNIT));
      const rm = teamRes[u.teamId];
      if (rm) {
        rm.gold += gold;
        rm.onChange?.();
      }
      route.active = false;
    }
  }

  private _nearestMarket(
    buildings: Building[],
    pos: THREE.Vector3,
    teamId: number,
    ownTeam: boolean,
  ): Building | null {
    let best: Building | null = null;
    let bestDist = Infinity;
    for (const b of buildings) {
      if (!b.alive || b.buildingType !== BuildingType.Market) continue;
      if (ownTeam && b.teamId !== teamId) continue;
      if (!ownTeam && b.teamId === teamId) continue;
      const d = pos.distanceTo(b.pos);
      if (d < bestDist) { bestDist = d; best = b; }
    }
    return best;
  }

  /** Remove dead or deselected carts from state. */
  prune(units: Unit[]): void {
    for (const u of this._routes.keys()) {
      if (!units.includes(u) || !u.alive) this._routes.delete(u);
    }
  }
}
